mhs-reference-implementation/mhs/parser/ebxml_message_parser.py

```diff
@@ -1,38 +1,137 @@
+import email
+import email.policy
+import logging
+import xml.etree.ElementTree as ET
+
 FROM_PARTY_ID = "from_party_id"
+TO_PARTY_ID = "to_party_id"
 CPA_ID = "cpa_id"
 CONVERSATION_ID = "conversation_id"
+SERVICE = "service"
+ACTION = "action"
 MESSAGE_ID = "message_id"
 TIMESTAMP = "timestamp"
+REF_TO_MESSAGE_ID = "ref_to_message_id"
+DUPLICATE_ELIMINATION = "duplicate_elimination"
+ACK_REQUESTED = "ack_requested"
+ACK_SOAP_ACTOR = "ack_soap_actor"
+SYNC_REPLY = "sync_reply"
+
+MESSAGE = "message"
+
+EBXML_NAMESPACE = "eb"
+SOAP_NAMESPACE = "SOAP"
+NAMESPACES = {SOAP_NAMESPACE: "http://schemas.xmlsoap.org/soap/envelope/",
+              EBXML_NAMESPACE: "http://www.oasis-open.org/committees/ebxml-msg/schema/msg-header-2_0.xsd"}
+
+CONTENT_TYPE_HEADER_NAME = "Content-Type"
+
+
+class EbXmlParsingError(Exception):
+    """Raised when an error was encountered during parsing of an ebXML message."""
+    pass
 
 
 class EbXmlMessageParser:
     """A component that extracts information from ebXML messages."""
 
-    def parse_message(self, message):
+    def parse_message(self, headers, message):
         """Parse the provided ebXML message and extract a dictionary of values from it.
 
+        :param headers A dictionary of headers received with the message.
         :param message: The message to be parsed.
         :return: A dictionary of values extracted from the message.
         """
-        # TODO: Implement schematron/XSLT parsing here.
-        pass
+        xml_tree = ET.fromstring(message)
+        extracted_values = {}
 
+        self._extract_ebxml_text_value(xml_tree, "PartyId", extracted_values, FROM_PARTY_ID, parent="From")
+        self._extract_ebxml_text_value(xml_tree, "PartyId", extracted_values, TO_PARTY_ID, parent="To")
+        self._extract_ebxml_text_value(xml_tree, "CPAId", extracted_values, CPA_ID)
+        self._extract_ebxml_text_value(xml_tree, "ConversationId", extracted_values, CONVERSATION_ID)
+        self._extract_ebxml_text_value(xml_tree, "Service", extracted_values, SERVICE)
+        self._extract_ebxml_text_value(xml_tree, "Action", extracted_values, ACTION)
+        self._extract_ebxml_text_value(xml_tree, "MessageId", extracted_values, MESSAGE_ID, parent="MessageData")
+        self._extract_ebxml_text_value(xml_tree, "Timestamp", extracted_values, TIMESTAMP, parent="MessageData")
+        self._extract_ebxml_text_value(xml_tree, "RefToMessageId", extracted_values, REF_TO_MESSAGE_ID,
+                                       parent="MessageData")
```

Comment:
Would it work to have the `xml_tree` set as a member variable here, seems like we're passing that tree round a lot of methods and it could save a parameter

Reply:
My intention was that a single instance of `EbXmlMessageParser` could be used to parse multiple messages, so setting this tree as a member variable wouldn't make sense in that respect.

Reply:
Makes sense, though if the case is this class doesn't retain state it makes me wonder if it should be static but I don't think this would add any value